src/v/cluster/isolation_node_watcher.cc

```diff
@@ -0,0 +1,110 @@
+  ss::sharded<node_status_table>& node_status_table)
+  : _metadata_cache(metadata_cache)
+  , _health_monitor(health_monitor)
+  , _node_status_table(node_status_table) {
+    start_isolation_watch_timer();
+}
+
+ss::future<> isolation_node_watcher::stop() {
+    _isolation_watch_timer.cancel();
+    return _gate.close();
+}
+
+void isolation_node_watcher::start_isolation_watch_timer() {
+    _isolation_watch_timer.set_callback([this] { update_isolation_status(); });
+    rearm_isolation_watch_timer();
+}
+
+void isolation_node_watcher::rearm_isolation_watch_timer() {
+    _isolation_watch_timer.arm(
+      model::timeout_clock::now() + _isolation_check_ms);
+}
+
+void isolation_node_watcher::update_isolation_status() {
+    ssx::spawn_with_gate(
+      _gate, [this] { return do_update_isolation_status(); });
+}
+
+ss::future<> isolation_node_watcher::do_update_isolation_status() {
+    bool isolated_status = co_await is_node_isolated();
+    if (isolated_status != _last_is_isolated_status) {
+        vlog(
+          clusterlog.debug,
+          "Change is_isolated status. Is node isolated: {}",
+          isolated_status);
+        co_await _metadata_cache.invoke_on_all(
+          [isolated_status](cluster::metadata_cache& mc) {
+              mc.set_is_node_isolated_status(isolated_status);
+          });
+        _last_is_isolated_status = isolated_status;
+    }
+    rearm_isolation_watch_timer();
+}
+
+ss::future<bool> isolation_node_watcher::is_node_isolated() {
+    auto nodes_status_res = co_await _health_monitor.local().get_nodes_status(
+      config::shard_local_cfg().metadata_status_wait_timeout_ms()
+      + model::timeout_clock::now());
+
+    if (nodes_status_res.has_value()) {
+        auto& nodes_status = nodes_status_res.value();
+
+        // All nodes from health report should be not alive for isolation
+        if (std::any_of(
+              nodes_status.begin(),
+              nodes_status.end(),
+              [](const node_state& node_state) {
+                  // Current node should not influence on answer.
+                  if (config::node().node_id() == node_state.id) {
+                      return cluster::alive::no;
+                  }
+                  return node_state.is_alive;
+              })) {
+            co_return false;
+        }
+    }
+
+    auto controller_leader_id
+      = _metadata_cache.local().get_controller_leader_id();
+    // Check last append_etries request time makes sense only for follower
+    if (
+      controller_leader_id.has_value()
+      && config::node().node_id() != controller_leader_id.value()) {
+        bool is_last_append_entries_request_expired
+          = co_await _health_monitor.local()
+              .is_controller_last_append_entries_request_expired();
+        if (!is_last_append_entries_request_expired) {
+            co_return false;
+        }
+    }
+
+    if (!_node_status_table.local().is_isolated()) {
```

Comment:
maybe it would be better to do this cheap check first and only then proceed with more exspensive call outs to health monitor ?